feat(firebase_crashlytics): rework

Salakar opened the pull request:
---

Co-authored-by: @MaikuB 
Co-authored-by: @Moncader
Co-authored-by: @axel-op
Co-authored-by: @Ehesp
Co-authored-by: @Salakar 
Co-authored-by: @greghesp 
Co-authored-by: @kirstywilliams
Co-authored-by: @helenaford 

---

## Description

For testing purposes, `firebase_crashlytics` version `0.2.0-dev.4` is published with these changes.

### Changes

- **BREAKING**: Removal of Fabric SDKs and migration to the new Firebase Crashlytics SDK.
- **BREAKING**: The following methods have been removed as they are no longer available on the Firebase Crashlytics SDK:
  - `setUserEmail`
  - `setUserName`
  - `getVersion`
  - `isDebuggable`
- **BREAKING**: `log` now returns a Future. Calling `log` now sends logs immediately to the underlying Crashlytics SDK instead of pooling them in Dart.
- **BREAKING**: the methods `setInt`, `setDouble`, `setString` and `setBool` have been replaced by `setCustomKey`.
  - `setCustomKey` returns a Future. Calling `setCustomKey` now sends custom keys immediately to the underlying Crashlytics SDK instead of pooling them in Dart.
- **DEPRECATED**: `enableInDevMode` has been deprecated, use `isCrashlyticsCollectionEnabled` and `setCrashlyticsCollectionEnabled` instead.
- **DEPRECATED**: `Crashlytics` has been deprecated, use `FirebaseCrashlytics` instead.
- **NEW**: Custom keys that are automatically added by FlutterFire when calling `reportError` are now prefixed with `flutter_error_`.
- **NEW**: Calling `.crash()` on Android  & iOS/macOS now reports a custom named exception to the Firebase Console. This allows you to easily locate test crashes.
  - Name: `FirebaseCrashlyticsTestCrash`.
  - Message: `This is a test crash caused by calling .crash() in Dart.`.
- **NEW**: `recordError` now uses a named native exception when reporting to the Firebase Console. This allows you to easily locate errors originating from Flutter.
  - Name: `FlutterError`.
- **NEW**: Added support for `checkForUnsentReports`.
  - Checks a device for any fatal or non-fatal crash reports that haven't yet been sent to Crashlytics.
  - See reference API docs for more information.
- **NEW**: Added support for `deleteUnsentReports`.
  - If automatic data collection is disabled, this method queues up all the reports on a device for deletion.
  - See reference API docs for more information.
- **NEW**: Added support for `didCrashOnPreviousExecution`.
  - Checks whether the app crashed on its previous run.
  - See reference API docs for more information.
- **NEW**: Added support for `sendUnsentReports`.
  - If automatic data collection is disabled, this method queues up all the reports on a device to send to Crashlytics.
  - See reference API docs for more information.
- **NEW**: Added support for `setCrashlyticsCollectionEnabled`.
  - Enables/disables automatic data collection by Crashlytics.
  - See reference API docs for more information.
- **NEW**: Added support for `isCrashlyticsCollectionEnabled`.
  - Whether the current Crashlytics instance is collecting reports. If false, then no crash reporting data is sent to Firebase.
  - See reference API docs for more information.
- **FIX**: Fixed a bug that prevented keys from being set on iOS devices.


## Related Issues

Supersedes #2288 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] If the pull request affects only one plugin, the PR title starts with the name of the plugin in brackets (e.g. [cloud_firestore])
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [ ] No, this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

Salakar commented:
Tagging in everyone from the previous PR for a final review pass:

 - @MaikuB
 - @Moncader
 - @axel-op
 - @Ehesp
 - @Salakar
 - @greghesp
 - @kirstywilliams
 - @helenaford

Salakar commented:
If it helps testing this out before we merge, I can publish a dev release. Let me know.

HofmannZ commented:
@Salakar a dev release would help ✌🏼

Salakar commented:
@HofmannZ `0.2.0-dev.1` is up

tillchen commented:
👍

HofmannZ commented:
@Salakar it seems like `enableInDevMode` on is not exposed on `FirebaseCrashlytics.instance`. Therfore `FirebaseCrashlytics.instance.enableInDevMode = true;` from the useage docs does not work.

<img width="1669" alt="Screenshot 2020-09-01 at 11 56 16" src="https://user-images.githubusercontent.com/17142193/91835795-2888a400-ec4a-11ea-86ac-28a44dfdfd8d.png">

Salakar commented:
> @Salakar it seems like `enableInDevMode` on is not exposed on `FirebaseCrashlytics.instance`. Therfore `FirebaseCrashlytics.instance.enableInDevMode = true;` from the useage docs does not work.
> 
> <img alt="Screenshot 2020-09-01 at 11 56 16" width="1669" src="https://user-images.githubusercontent.com/17142193/91835795-2888a400-ec4a-11ea-86ac-28a44dfdfd8d.png">

Docs need changing here, should be using `isCrashlyticsCollectionEnabled` and `setCrashlyticsCollectionEnabled(true/false)` now - this now toggles crash reporting natively,  so when its off now both Dart and native crash reporting are disabled. Previously `enableInDevMode` only toggled reports from Dart, any errors happening natively would still have been reported.

You could set this programatically in your code, e.g.:

```dart
if (kDebugMode) {
  // Force enable crashlytics collection enabled if we're testing it.
  await FirebaseCrashlytics.instance
      .setCrashlyticsCollectionEnabled(true);
}
```

Hope that makes sense?

HofmannZ commented:
Android works fine, having some issues with iOS pods. Let me check if that's me or the package.

HofmannZ commented:
Ok, resolved the pods issue by cleaning the pods: `rm -rf ios/Pods && rm ios/Podfile && rm ios/Podfile.lock`.

But running into the following issue during iOS the build:
```
duplicate symbol '_kArgumentCode' in:
    /Users/zinohofmann/Library/Developer/Xcode/DerivedData/Runner-aefewpetgecaciadgvxxfimlevof/Build/Products/Debug-dev-iphonesimulator/firebase_auth/firebase_auth.framework/firebase_auth(FLTFirebaseAuthPlugin.o)
    /Users/zinohofmann/Library/Developer/Xcode/DerivedData/Runner-aefewpetgecaciadgvxxfimlevof/Build/Products/Debug-dev-iphonesimulator/firebase_crashlytics/firebase_crashlytics.framework/firebase_crashlytics(FLTFirebaseCrashlyticsPlugin.o)
ld: 1 duplicate symbol for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Salakar commented:
Interesting, I know what that is, will push a fix shortly

Salakar commented:
Have pushed new docs for the `enabledInDevMode` replacements, I've also added `enabledInDevMode` back in but with deprecation warnings so that it's not a hard deprecation

MaikuB commented on docs/crashlytics/overview.mdx:
```diff
@@ -0,0 +1,83 @@
+dependencies:
+  flutter:
+    sdk: flutter
+  firebase_core: "^{{ plugins.firebase_core }}"
```
Haven't been following the updates to `firebase_core` so I may have missed something but is this step needed as won't it get pulled in as a transitive dependency?

HofmannZ commented:
> Have pushed new docs for the `enabledInDevMode` replacements, I've also added `enabledInDevMode` back in but with deprecation warnings so that it's not a hard deprecation

Since crashlytics collection is enabled by default, it might be good to explain it's a best practice to disable it during development.

```dart
  if (kDebugMode) {
    // Force disable crashlytics collection enabled while doing every day development.
    await FirebaseCrashlytics.instance.setCrashlyticsCollectionEnabled(false);
  }
```

HofmannZ commented:
Also, the check `Deploy Preview / build_and_preview` silently failed because it `Cannot find module './crashlytics-filter.png'`.

Salakar commented:
> Also, the check `Deploy Preview / build_and_preview` silently failed because it `Cannot find module './crashlytics-filter.png'`.

Aye, have seen this, am looking into it :) thanks

MaikuB commented on docs/crashlytics/overview.mdx:
````diff
@@ -0,0 +1,83 @@
+
+1. Add the following classpaths to the `/android/build.gradle` file.
+
+```dart
````
Shouldn't this be _```gradle_?

MaikuB commented on docs/crashlytics/usage.mdx:
````diff
@@ -1,6 +1,171 @@
+import 'package:firebase_crashlytics/firebase_crashlytics.dart';
+```
+
+## Sending reports to Crashlytics
````
nit: may want to choose being consistent on using title case (e.g. see reports.mdx) vs sentence casing

HofmannZ commented on docs/crashlytics/usage.mdx:
```diff
@@ -18,17 +18,18 @@ to Firebase the next time the application is launched.
+  // Temporarily toggle this to true if you want to test crash reporting in your app.
   await FirebaseCrashlytics.instance
-      .setCrashlyticsCollectionEnabled(true);
+      .setCrashlyticsCollectionEnabled(false);
```
✅

MaikuB commented on packages/firebase_crashlytics/firebase_crashlytics/CHANGELOG.md:
```diff
@@ -1,3 +1,15 @@
+## 0.2.0-dev.1
+
+* **Breaking change**: Migration to the new Firebase Crashlytics SDK. Follow the [migration guide](https://firebase.google.com/docs/crashlytics/upgrade-sdk?platform=android) (steps 1 and 2) to update your app.
```
Would it perhaps be better to point to the location of the docs for this plugin? I notice from the link that it points to Android so the thing that comes to mind as a reader is does this only affect Android and what about iOS

Salakar replied:
Specific steps to migrate are now covered explicitly in the migration guide:

![image](https://user-images.githubusercontent.com/5347038/91998275-ed15d480-ed32-11ea-8557-b2fa85d5ec3e.png)

MaikuB commented on packages/firebase_crashlytics/firebase_crashlytics/example/android/app/build.gradle:
```diff
@@ -32,10 +32,9 @@ android {
+        applicationId 'io.flutter.plugins.firebasecrashlyticsexample'
         minSdkVersion 16
-        targetSdkVersion 27
+        targetSdkVersion 28
```
if this is getting bumped , i wonder if it's worth bumping it to 29 even if it doesn't change much/anything

Salakar replied:
We'll be bumping all example apps separately to 30, already tracking this - thanks :)

MaikuB replied:
Np, I guess a similar comment applies to the compile SDK version specified for the plugin

MaikuB commented on packages/firebase_crashlytics/firebase_crashlytics/CHANGELOG.md:
```diff
@@ -1,3 +1,15 @@
+  * `isDebuggable`
+* **Breaking change**: the methods `setInt`, `setDouble`, `setString` and `setBool` have been replaced by `setCustomKey`.
+* Logs and keys are sent to Crashlytics as they're set.
+* Fixes a bug that prevented keys from being set on iOS devices.
```
Should the changelog also add mention of bumping the Gradle plugin used? I've seen this documented in the changelogs for other first-party plugins before

Salakar replied:
Covered by the migration guide. Added a link in the changelog to point to the migration guide.

helenaford commented on packages/firebase_crashlytics/firebase_crashlytics_platform_interface/README.md:
```diff
@@ -0,0 +1,26 @@
+
+A common platform interface for the [`firebase_crashlytics`][1] plugin.
+
+This interface allows platform-specific implementations of the `cloud_firestore`
```
```suggestion
This interface allows platform-specific implementations of the `firebase_crashlytics`
```

helenaford commented on packages/firebase_crashlytics/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -0,0 +1,213 @@
+                duration: Duration(seconds: 5),
+              ));
+
+              // // Example of an exception that does not get caught
```
```suggestion
              // Example of an exception that does not get caught
```

helenaford commented on packages/firebase_crashlytics/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -0,0 +1,213 @@
+              ));
+
+              // // Example of an exception that does not get caught
+              // // by `FlutterError.onError` but is caught by
```
```suggestion
              // by `FlutterError.onError` but is caught by
```

helenaford commented on packages/firebase_crashlytics/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -0,0 +1,213 @@
+
+              // // Example of an exception that does not get caught
+              // // by `FlutterError.onError` but is caught by
+              // // `runZonedGuarded`.
```
```suggestion
              // `runZonedGuarded`.
```

MaikuB commented on packages/firebase_crashlytics/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -0,0 +1,213 @@
+
+  @override
+  void initState() {
+    initializeFlutterFire();
```
The convention afaik is to generally call the `super.initState` so wondering if there's a reason it's last in this case

MaikuB commented on packages/firebase_crashlytics/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -0,0 +1,213 @@
+      };
+
+      setState(() {
+        _initialized = true;
```
`initializeFlutterFire()` calls async code so unlikely to run into this scenario but if it was entirely synchronous, the code could fail when calling `setState` as the widget hasn't been mounted. May be worth putting a guard around the `setState` call  by checking if it's mounted and if it hasn't then could set the value of `_initialized` without wrapping with a `setState()` call

Salakar replied:
@MaikuB would you be willing to send a PR to https://github.com/invertase/flutterfire/tree/%40invertase/crashlytics_v1/packages to improve the example app with this point and all the other points you've mentioned?

MaikuB replied:
To clarify, you're referring to a PR for just the changes to the example app right? If so then yeah I can take a look. Will likely need help testing to make sure it's working though last time issues with getting the app to build on my machine. Hopefully have better luck this time...

Salakar replied:
Yes please, just the example app. We have tooling now for the monorepo that should help you get setup locally now without too much difficulty, updated contributing guide is here (wip, so not yet merged): https://github.com/FirebaseExtended/flutterfire/blob/@ehesp/update-contributing/CONTRIBUTING.md

If you're using an IntelliJ based IDE then the tooling will automatically generate run configurations for you so you can just run it from there directly;

![image](https://user-images.githubusercontent.com/5347038/91851405-00ec0880-ec57-11ea-8eb3-7791d698e654.png)

MaikuB replied:
Hven't done this before but how I would start making changes based on the fork your team has as I already have a fork based on this repo

MaikuB replied:
Actually nevermind, GitHub client lets me branch from this PR :)

MaikuB commented on packages/firebase_crashlytics/firebase_crashlytics/example/lib/main.dart:
```diff
@@ -0,0 +1,213 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return withMaterialApp(Center(child: appContent));
```
Preference is to split build logic into smaller widgets as opposed to methods (see https://iiro.dev/2018/12/11/splitting-widgets-to-methods-performance-antipattern/) so may want to consider updating the code here

helenaford commented on packages/firebase_crashlytics/firebase_crashlytics/example/android/app/src/main/java/io/flutter/plugins/firebasecrashlyticsexample/MainActivity.java:
```diff
@@ -1,16 +1,18 @@
+import io.flutter.plugins.firebase.crashlytics.FlutterFirebaseCrashlyticsPlugin;
 
 public class MainActivity extends FlutterActivity {
   // TODO(<github-username>): Remove this once v2 of GeneratedPluginRegistrant
```
Spotted a TODO. MainActivity needs to be updated in example app

Salakar commented:
@HofmannZ `0.2.0-dev.2` is up with iOS/macOS fixes

HofmannZ commented on docs/crashlytics/overview.mdx:
```diff
@@ -33,29 +33,28 @@ $ flutter pub get
-  implementation 'com.google.firebase:firebase-analytics:17.5.0'
-
-  // Add the Firebase Crashlytics SDK.
-  implementation 'com.google.firebase:firebase-crashlytics:17.2.1'
```
I think you removed a bit too much here.

Salakar replied:
These dependencies are managed by the plugin itself which includes them transitively, so its not required by the user to add this in

HofmannZ commented:
With `0.2.0-dev.2` iOS builds fine and the test `.crash()` is reported to the console. 🚀

Tho it has a warning that I should `Upload 1 missing dSYM required to process 1 crash` even though I have the `${PODS_ROOT}/FirebaseCrashlytics/run` build phase. That should be taken care of by the build phase, right?

krokyze commented:
It looks like there's an issue on iOS when using `crashlytics.log` - https://puu.sh/Go8Qt/1ee371b2c2.png 

On Android everything works as expected 🎉

Salakar commented:
> It looks like there's an issue on iOS when using `crashlytics.log` - https://puu.sh/Go8Qt/1ee371b2c2.png
> 
> On Android everything works as expected

Thanks, looking into this

HofmannZ commented:
Okay, I did a release build with CodeMagic, but still, no .dSYM files show up in the Firebase console.

For convenience, I named the build phase `Run FirebaseCrashlytics` and it's run after the dSYM files are generated:

<img width="450" alt="Screenshot 2020-09-01 at 19 44 09" src="https://user-images.githubusercontent.com/17142193/91887786-8689ab80-ec8b-11ea-9128-cf7048759bd9.png">

The only thing the build phase does is: 

```sh
${PODS_ROOT}/FirebaseCrashlytics/run
```

I also tried adding `$(BUILT_PRODUCTS_DIR)/$(INFOPLIST_PATH)` as an input file with the same result.

Salakar commented:
@HofmannZ Is there any additional log output showing in this section: 
![image](https://user-images.githubusercontent.com/5347038/91891227-8f2cb280-ec88-11ea-9dfa-691a69c1853d.png)

HofmannZ commented:
@Salakar Yes:

```
Running upload-symbols in Build Phase mode
Validating build environment for Crashlytics...
Validation succeeded. Exiting because upload-symbols was run in validation mode
```

I actually found that the `${PODS_ROOT}/FirebaseCrashlytics/run` script is executed in the background as per docs:

```
# This script is meant to be run as a Run Script in the "Build Phases" section
# of your Xcode project. It sends debug symbols to symbolicate stacktraces,
# sends build events to track versions, and onboards apps for Crashlytics.
#
# This script calls upload-symbols twice:
#
# 1) First it calls upload-symbols synchronously in "validation" mode. If the
#    script finds issues with the build environment, it will report errors to Xcode.
#    In validation mode it exits before doing any time consuming work.
#
# 2) Then it calls upload-symbols in the background to actually send the build
#    event and upload symbols. It does this in the background so that it doesn't
#    slow down your builds. If an error happens here, you won't see it in Xcode.
#
# You can find the output for the background execution in Console.app, by
# searching for "upload-symbols".
#
```

I think that the build container exits to early for the actual upload to happen. I'm testing a new build where I replaced the command in the build phase with:

```sh
${PODS_ROOT}/FirebaseCrashlytics/upload-symbols --build-phase
```

This should force the container to stay running until the upload is done. If this works we should probably mention it in the docs.

Salakar commented:
> You can find the output for the background execution in Console.app, by searching for "upload-symbols".

Probably worth checking these logs on your Mac to see if it had any errors uploading

HofmannZ commented:
That worked, the output is now:

```
Running upload-symbols in Build Phase mode
Validating build environment for Crashlytics...
Fetching upload-symbols settings...
Sending Crashlytics build event...
Processing dSYMs...
Successfully submitted symbols for architecture armv7 with UUID 153e122cfe6c38beb24424d3041def9a in dSYM: /Users/builder/Library/Developer/Xcode/DerivedData/Runner-edaimyiflreloheqntgnhkmwcclv/Build/Intermediates.noindex/ArchiveIntermediates/dev/BuildProductsPath/Release-dev-iphoneos/Runner.app.dSYM
Successfully submitted symbols for architecture arm64 with UUID a91f01cb0fae302f9aca7defff81b8de in dSYM: /Users/builder/Library/Developer/Xcode/DerivedData/Runner-edaimyiflreloheqntgnhkmwcclv/Build/Intermediates.noindex/ArchiveIntermediates/dev/BuildProductsPath/Release-dev-iphoneos/Runner.app.dSYM
Successfully uploaded Crashlytics build event and symbols
```

Test `.crash()` is showing up in the Firebase console.

HofmannZ commented:
> Probably worth checking these logs on your Mac to see if it had any errors uploading. We could document this also. But ye if your build container exits early then we should probably document to use upload-symbols instead, we could do this based on if CI env exists then use upload-symbols, otherwise run as normal?

@Salakar I think that would indeed be the long term solution. For now, we should al least mention that if your building with a CI, you should use `${PODS_ROOT}/FirebaseCrashlytics/upload-symbols --build-phase` instead of `${PODS_ROOT}/FirebaseCrashlytics/run`.

krokyze commented:
Tried using `${PODS_ROOT}/FirebaseCrashlytics/upload-symbols --build-phase` and it breaks debug builds for me.

Build prints out bunch of warnings about pods and ends with:
```
upload-symbols[40295:224829] Unable to get file attributes for dSYM file at path
"/project/build/ios/Debug-iphoneos/Runner.app.dSYM/Contents/Resources/DWARF/Runner"
Running upload-symbols in Build Phase mode
Validating build environment for Crashlytics...
warning: DEBUG_INFORMATION_FORMAT should be set to dwarf-with-dsym for all configurations. This could also be a timing issue, make sure the Fabric run script build phase is the last build phase and no other scripts have moved the dSYM from the location Xcode generated it. Unable to process Runner.app.dSYM at path /project/build/ios/Debug-iphoneos/Runner.app.dSYM
Make sure your project build settings are generating a dSYM file.

Fetching upload-symbols settings...
Sending Crashlytics build event...
Processing dSYMs...
Command PhaseScriptExecution failed with a nonzero exit code
note: Using new build system
note: Building targets in parallel
note: Planning build
note: Constructing build description

Could not build the precompiled application for the device.
```

Salakar commented:
Hey all, `0.2.0-dev.3` is up with the following fixes:

* Fixed an issue with iOS logs appearing as `null` on the Firebase Console.
* Ensure cosmetic Android exception classes keep their name when app obfuscated.
* Fixed an error when hot restarting on Android (#3432).

HofmannZ commented:
@krokyze I've been digging into the issue a little bit. And it seems that the output for the background execution in `Console.app` actually always fails with that error. But since it's running in the background, you don't see the error.

The root of the problem lies in the fact that Flutter, in debug mode, does not generate dSYM files. If we move that out of the background the debug builds fail.

One solution would be to find a way to check of dev mode in the build phase. As for now, I have no idea how to do that. Any suggestions are welcome :)

krokyze commented:
@HofmannZ I was thinking the same about checking if it's debug mode, not sure if it's possible. The more I searched for solution, the more I started to think this is not really issue related with this package, but CI and FirebaseCrashlytics issue https://github.com/fastlane/fastlane/issues/13096

Salakar commented:
> @krokyze I've been digging into the issue a little bit. And it seems that the output for the background execution in `Console.app` actually always fails with that error. But since it's running in the background, you don't see the error.
> 
> The root of the problem lies in the fact that Flutter, in debug mode, does not generate dSYM files. If we move that out of the background the debug builds fail.
> 
> One solution would be to find a way to check for debug mode in the build phase. As for now, I have no idea how to do that. Any suggestions are welcome :)

You could do something like the following to limit to release builds only:

```
if [ "${CONFIGURATION}" = "Release" ]; then
  ${PODS_ROOT}/FirebaseCrashlytics/upload-symbols --build-phase
fi
```

HofmannZ commented:
> @HofmannZ I was thinking the same about checking if it's debug mode, not sure if it's possible. The more I searched for solution, the more I started to think this is not really issue related with this package, but CI and FirebaseCrashlytics issue [fastlane/fastlane#13096](https://github.com/fastlane/fastlane/issues/13096)

Well, with a native iOS app, you would use `dwarf-with-dsym` for all configurations. And therefore the upload would be all fine. Our specific issue comes from the fact that Flutter does not enable `dwarf-with-dsym` for the Debug configuration by default. One option would be to enable it for debug mode ourselves. But I think the solution pointed out by @Salakar might actually to the job as well and only requires two additional lines. In addition, I have no idea what the effect of enabling `dwarf-with-dsym` for the Debug would have on the live-reload and other Flutter internals.

Going to test it 🧪

Salakar commented:
Final script I've come up locally with, that will check if CI (supports all CIs) and whether release:


```bash
IS_CI="${CI}${CONTINUOUS_INTEGRATION}${BUILD_NUMBER}${RUN_ID}"
if [[ ("${CONFIGURATION}" == "Release" && (-n "${IS_CI}")) ]]; then
  ${PODS_ROOT}/FirebaseCrashlytics/upload-symbols --build-phase
else
  ${PODS_ROOT}/FirebaseCrashlytics/run
fi
```

HofmannZ commented:
> Final script I've come up locally with, that will check if CI (should support all CIs) and whether is release mode:
> 
> ```shell
> IS_CI="${CI}${CONTINUOUS_INTEGRATION}${BUILD_NUMBER}${RUN_ID}"
> if [[ ("${CONFIGURATION}" == "Release" && (-n "${IS_CI}")) ]]; then
>   ${PODS_ROOT}/FirebaseCrashlytics/upload-symbols --build-phase
> else
>   ${PODS_ROOT}/FirebaseCrashlytics/run
> fi
> ```

Don't we only want to run any of the scripts if in release mode, and then pick the script based on if you are in CI?

Salakar commented:
I think everything is generally good for a merge here, so I'll go ahead and do one final dev release off this branch and then merge.

We should do some follow up PRs for things like docs (e.g. the run script in CI notes from above) - would welcome contributions for this if anyone is willing.

We'll keep it on dev release for a couple of weeks.
